url29ecd5d728_geodb_db3dcore_tgz-pJ8-de_uos_igf_db3d_dbms_model3d_SegmentNetBuilderJ8/src/de/uos/igf/db3d/dbms/util/TetrahedronServices.java

```diff
@@ -0,0 +1,354 @@
+		return pointIDs;
+	}
+
+	/**
+	 * The Map "tetrahedrons" is a Map of TetrahedronIDs with an Integer Array
+	 * of 4 PointIDs.
+	 * 
+	 * @return the triangles
+	 */
+	public HashMap<Integer, int[]> getTetrahedrons() {
+		return tetrahedrons;
+	}
+
+	/**
+	 * This Map contains all componentIDs accessible by their tetrahedronID
+	 * (tetrahedronID, componentID)
+	 * 
+	 * @return the components
+	 */
+	public HashMap<Integer, Integer> getComponents() {
+		return components;
+	}
+
+	/**
+	 * This Map contains the attributes from the tetrahedron net moved to RGB
+	 * (pointID, color)
+	 * 
+	 * @return the components
+	 */
+	public HashMap<Integer, String> getAttributeAsColor(String attribute) {
+
+		HashMap<Integer, String> attributeColors = new HashMap<Integer, String>();
+
+		double maxValue = Double.MIN_VALUE;
+		double minValue = Double.MAX_VALUE;
+
+		// check out the range of values:
+		for (Point3D point : points.values()) {
+			double tmp = Double.valueOf(point.getAttributeValue(attribute));
+			if (tmp > maxValue)
+				maxValue = tmp;
+			else if (tmp < minValue)
+				minValue = tmp;
+		}
+
+		minValue = Math.abs(minValue);
+		maxValue = maxValue + minValue;
+
+		// build colors:
+		for (Point3D point : points.values()) {
+
+			double tmp = Double.valueOf(point.getAttributeValue(attribute));
+			tmp = tmp + minValue;
+			tmp = (510. * tmp) / maxValue;
+			String hex, hexTmp;
+
+			if (tmp <= 255) {
+				hexTmp = Integer.toHexString((int) tmp);
+				if (hexTmp.length() == 1)
+					hexTmp = "0" + hexTmp;
+				hex = hexTmp + "ff";
+			} else {
+				hexTmp = Integer.toHexString((int) tmp - 255);
+				if (hexTmp.length() == 1)
+					hexTmp = "0" + hexTmp;
+				hex = "ff" + hexTmp;
+			}
+
+			attributeColors.put(pointIDs.get(point), "0xff" + hex + "00");
+		}
+		return attributeColors;
+	}
+
+	/**
+	 * Converts RGB values to hex!
+	 * 
+	 * @param r
+	 * @param g
+	 * @param b
+	 * @return
+	 */
+	private static String colorToHex(String r, String g, String b) {
+		String rHex = Integer.toHexString(Integer.parseInt(r));
+		String gHex = Integer.toHexString(Integer.parseInt(g));
+		String bHex = Integer.toHexString(Integer.parseInt(b));
+
+		return ("0xff" + rHex + gHex + bHex);
+	}
+
+	public void createBoundaryElements(TetrahedronNet4D net,
+			Map<Integer, Point3D> pointTubes4D, Date date) {		
+
+		// save the unique Segments of this TetrahedronNet
+		Map<Segment4D, Integer> uniqueSegments = new HashMap<Segment4D, Integer>();
+		
+		// save the unique Triangles of this TetrahedronNet
+		Map<Triangle4D, Integer> uniqueTriangles = new HashMap<Triangle4D, Integer>();
+		
+		// first segID
+		int segID = net.getBoundaryElements1D(date).size();
+
+		// first triID
+		int triID = net.getBoundaryElements2D(date).size();
+		
+		Map<Integer, Element4D> tetrahedronElements = net.getNetElements(date);
+
+		for (Integer tetrahedronID : tetrahedronElements.keySet()) {
+
+			Tetrahedron4D tmpTetrahedron = (Tetrahedron4D) tetrahedronElements
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.